foyer/general_forcefield.py

```diff
@@ -0,0 +1,417 @@
+        return None
+
+
+    def apply(self, top, references_file=None, use_residue_map=True,
+    assert_bond_params=True, assert_angle_params=True,
+    assert_dihedral_params=True, assert_improper_params=True,
+    verbose=False, *args, **kwargs):
+        """Apply the force field to a molecular topology
+
+        Parameters
+        ----------
+        top : gmso.Topology, or mb.Compound
+            Molecular Topology to apply the force field to
+        references_file : str, optional, defaut=None
+            Name of file where force field refrences will be written (in Bibtex format).
+        use_residue_map : bool, optional, default=True
+            Options to speed up if there are a lot of repeated
+            subtopology within the topology (assuming they all
+            have the same name).
+        assert_bond_params : bool, optional, default=True
+            If True, Foyer will exit if parameters are not found for all system bonds.
+        assert_angle_params : bool, optional, default=True
+            If True, Foyer will exit if parameters are not found for all system angles.
+        assert_dihedral_params : bool, optional, default=True
+            If True, Foyer will exit if parameters are not found for all system dihedrals.
+        assert_improper_params : bool, optional, default=True
+            If True, Foyer will exit if parametes are not found for all system impropers.
+        verbose : bool, optional, default=False
+            If True, Foyer will print debug-level information about notable or potential problematic details it encounters.
+        """
+        if self.atomTypeDefinitions == {}:
+            raise FoyerError('Attempting to atom-type using a forcefield '
+                'with no atom type definitions.')
+
+        typemap = self._run_atomtyping(top,
+                    use_residue_map=use_residue_map,
+                    **kwargs)
+
+        return self._parametrize(top=top, typemap=typemap,
+            references_file=references_file,
+            assert_bond_params=assert_bond_params,
+            assert_angle_params=assert_angle_params,
+            assert_dihedral_params=assert_dihedral_params,
+            assert_improper_params=assert_improper_params,
+            verbose=verbose,
+            *args, **kwargs)
+
+    def _run_atomtyping(self, top, use_residue_map=True, **kwargs):
+        """Atomtype the topology
+
+        Parameters
+        ----------
+        top : gmso.Topology
+            Molecular Topology to find atom types for.
+        use_residue_map : bool, optional, default=True
+            Spped up options for duplicates subtopology.
+        """
+
+        if use_residue_map:
+            # Detect duplicates subtopology/residues
+            # (do matching by name, assert same number
+            # of atoms)
+            typemap = find_atomtypes(top, forcefield=self)
+        else:
+            typemap = find_atomtypes(top, forcefield=self)
```

Comment:
I know we used dictionaries as residue maps before, but I'm wondering if you could use [functools LRU caching](https://docs.python.org/3/library/functools.html), where the inputs to the function are some molecule's networkx graph, and the LRU caching means sometimes the function doesn't need to re-run if it's already cached the results from a previous function all

Reply:
Thank you! I will look into that